### PR TITLE
Secrets handling

### PR DESCRIPTION
**Pull request by dalehamel**
@wvanbergen @thegedge @xthexder @graemej for review

This does a couple of things related to secrets and variables:
- You can specify multiple secrets files. The default is to look for a config/secrets.ejson and a system path for /usr/local/cloudshaper/secrets.ejson (the latter is for shipit). This prevents duplicating the secret decryption logic, and simplifies my chef PR for shared provider secrets
- Secrets may contain multiple types of secrets, to reflect the reality of secrets in terraform
  - Terraform providers usually accept a specific environment variable for their configuration totally transparently, so we have 'providers' secrets for this
  - Terraform variables begin with TF_VAR_, and so we have 'variables' for this. 
- Secrets within ejson are now nested inside of the 'cloudshaper' hash, as to not collide with other application secrets.

### CODE REVIEW

**Comment by dalehamel** [comment]
Notably, this is completely devoid of testing. I promise i'll add tests later :trollface:

File: lib/cloudshaper/cli.rb [hunk 1]
```diff
@@ -4,6 +4,7 @@
 module Cloudshaper
   class CLI < Thor
     class_option 'remote_state', type: 'boolean'
+    class_option 'terraform_path', type: "string", default: 'terraform'
```
Comment (thegedge) [inline lib/cloudshaper/cli.rb]: Maybe a small description mentioning that this is the path to the binary (not the path containing the binary)?

File: lib/cloudshaper/secrets.rb [hunk 2]
```diff
@@ -1,29 +1,18 @@
...
-      access_key_id: 'foo',
-      secret_access_key: 'bar'
-    }
-  }
-else
-  SECRETS ||= begin
-    secrets_file = File.expand_path(ENV['CONFIG_PATH'] || 'config/secrets.json')
-    if File.exist?(secrets_file)
-      JSON.parse(File.read(secrets_file), symbolize_names: true, object_class: SecretHash)
+secrets_files = (ENV['SECRETS_FILES'] || '/usr/local/cloudshaper/secrets.ejson,config/secrets.ejson').split(',')
```
Comment (thegedge) [inline lib/cloudshaper/secrets.rb]: Maybe merge `ENV['SECRETS_FILES']` with the default paths?

File: lib/cloudshaper/command.rb [hunk 3]
```diff
@@ -10,18 +16,24 @@ def initialize(stack, command)
...
+            end
+          end
+        when 'variables'
+          secrets.each do |key, value|
+            @stack.variables[key.to_s] = v
           end
         end
       end
-      vars
+      @stack.variables.each { |k, v| environment["TF_VAR_#{k.to_s}"] = v }
```
Comment (thegedge) [inline lib/cloudshaper/command.rb]: Shell escape `v.to_s`.

**Comment by thegedge** [comment]
Other than a few small comments, LGTM!

File: lib/cloudshaper/cli.rb [hunk 4]
```diff
@@ -77,7 +78,8 @@ def version
 
     private
 
-    def load_stack(stack, _options, pull = true)
+    def load_stack(stack, pull = true)
```
Comment (wvanbergen) [inline lib/cloudshaper/cli.rb]: Named arguments is a lot nicer: `def load_stack(stack, pull: true)`
This allows you to call this method like this `load_stack(name, pull: false)`. 

This makes the intent of the boolean argument a lot clearer.

File: lib/cloudshaper/secrets.rb (hunk 2 above)
Comment (wvanbergen) [inline lib/cloudshaper/secrets.rb]: Also, let's make the default a constant.

File: lib/cloudshaper/secrets.rb [hunk 5]
```diff
@@ -1,29 +1,18 @@
 require 'json'
 require 'open3'
 
-# Load and provide access to secrets required by terraform providers
-class SecretHash < Hash
-  class SecretNotFound < Exception; end
+SECRETS ||= {}
```
Comment (wvanbergen) [inline lib/cloudshaper/secrets.rb]: I still think this should be `Cloudshaper::SECRETS` so we don't pollute the global namespace.

Comment (wvanbergen) [inline lib/cloudshaper/secrets.rb]: Why is it `||=`? Can this be set somewhere else already? That feels like a code smell given that it is supposed to be a constant.

File: lib/cloudshaper/secrets.rb [hunk 6]
```diff
@@ -1,29 +1,18 @@
...
+    if secrets_file.end_with?('.ejson')
+      secrets = `ejson decrypt #{secrets_file}`
+    elsif secrets_file.end_with('.json')
+      secrets = File.read(secrets_file)
     else
-      {}
+      fail "I don't understand how to get secrets from #{secrets_file}"
     end
+    SECRETS.merge!(JSON.parse(secrets)['cloudshaper'] || {})
   end
```
Comment (wvanbergen) [inline lib/cloudshaper/secrets.rb]:
``` ruby
module Cloudshaper
  SECRETS = secrets_files.inject({}) do |secrets, file|
    ...
    secrets.merge(JSON.parse(secret_blob))
  end.freeze
end
```

**Pull request merged**